### ISSUE TITLE
[Discussion] Improve checksum for git branch based spec

### ISSUE THREAD

**Issue by allenhsu** [CocoaPods/CocoaPods#9962]
* [x] I've read and understood the [*CONTRIBUTING* guidelines and have done my best effort to follow](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md).

# Report

## What did you do?

This issue was originally posted here https://github.com/CocoaPods/Core/issues/640 However there were no feedback for some time. So I repost it in this main repo. Sorry if it bothers.

Run `pod install` after switching branch.

## What did you expect to happen?

Install all pod dependencies correctly.

## What happened instead?

Installation is skipped for some pod, so incorrect cached version is used.

### Stack

```
   CocoaPods : 1.9.3
        Ruby : ruby 2.4.4p296 (2018-03-28 revision 63013) [x86_64-darwin19]
    RubyGems : 2.6.14.1
        Host : Mac OS X 10.15.6 (19G73)
       Xcode : 11.5 (11E608c)
         Git : git version 2.28.0
Ruby lib dir : /Users/allen/.rbenv/versions/2.4.4/lib
Repositories : trunk - CDN - https://cdn.cocoapods.org/
               upwlabs - git - git@github.com:upwlabs/PodsSpecs.git @ cf2344bbeb127b63f3159afaa66ad1ea8f4ade89

```

### Installation Source

```
Executable Path: /Users/allen/.rbenv/versions/2.4.4/bin/pod
```

### Plugins

```
claide-plugins         : 0.9.2
cocoapods-deintegrate  : 1.0.4
cocoapods-dependencies : 1.0.0.beta.1
cocoapods-plugins      : 1.0.0
cocoapods-search       : 1.0.0
cocoapods-stats        : 1.1.0
cocoapods-trunk        : 1.5.0
cocoapods-try          : 1.2.0
cocoapods_debug        : 0.1.0
```

## Project that demonstrates the issue

I created [this minimal project](https://github.com/allenhsu/cocoapods-checksum-demo) to reproduce the issue.

## Original Post

TL;DR, [current checksum is based on the spec file only](https://github.com/CocoaPods/Core/blob/59eaa9b1b22070d6c5ffcfa19f491c1f47594459/lib/cocoapods-core/specification.rb#L668), which may potentially cause wrong version of pod being used when using branch as source.

## Background

First of all, let me describe our development flow of private pods.

During development, we will check out the source code of the private pod into a local directory, checkout a new feature branch, e.g. `feature-name`, and point the pod to a local path using `pod "pod_name", :path => "local_path"`.

After local development, we will submit beta builds (of the main App containing the private pod) for QA to test. In this stage, we will use `pod "pod_name", :git => "source_repo", :branch => "feature-name"` to reference the latest code of the feature branch. Beta builds are build and deployed on a stateful CI server.

Only when it's stable for submission, we will then bump the version, make a tag, and push it to our private specs repo.

## Steps to Reproduce The Problem

I created [this minimal project](https://github.com/allenhsu/cocoapods-checksum-demo) to reproduce the issue.

1. Clone the repo, check out to `branch-based-pod`, run `pod install`, which will install `AFNetworking` from it's master branch.
2. Check out to `master` branch, run `pod install`, which is supposed to install the `AFNetworking` v4.0.1 as locked in Podfile.lock from the official trunk. However, because the checksum of `AFNetworking` didn't change, so pod skipped reinstalling it.

If you do it backward, there's no such issue, because branch based pod is pre-downloaded, which will hit [the check here](https://github.com/CocoaPods/CocoaPods/blob/7033bd7bb07a37041bc6e367b45660c9ac4784b8/lib/cocoapods/installer/analyzer/sandbox_analyzer.rb#L142).

This is causing some corner issues for us because we are using a local stateful CI server, Pods directory is not cleaned for each build. When switching to branches based on older release branches, the wrong cached pod may be used.

## How to Fix It

The quickest workaround for us is remove Pods directory before each build to make sure pods are installed from source, but this will make the build time longer than expected. For deployment tasks, we do remove Pods directory to make sure everything is clean. But for beta builds, we still want to keep it to accelerate the builds.

The better solution would be change how checksum is calculated, or how sandbox manifest and dependencies are compared. The checksum part is in this Core repo, while the sandbox analyzer is actually in the CocoaPods repo.

Not sure if anyone else has encountered similar problems, but I want to bring it up to see if there could be any improvements.

**Comment by dnkoutso** [comment]
thanks for filing it. It was discussed a bit in the CocoaPods Slack and PR opened here https://github.com/CocoaPods/CocoaPods/pull/9958

I am going to close the Core one to avoid duplication.

**Comment by allenhsu** [comment]
Thanks @dnkoutso. Good to know that it has already been taken care of.
